feat: webrtc m140 support

revidee opened the pull request:
note: this includes a similar fix as the PR336 (https://github.com/versatica/mediasoup-client/pull/336) from the mediasoup-client repo which deals with webrtc changes and expectations for payload-type/extension ids to stay consistent
note: some aestetics and clang-compile warnings have also been fixes, so it compiles using cpp20/clang20

I have to admit that no unit tests have been run, however, using it to produce/consume media works as expected in local tests.
This should also support M142, however, the latest I've tested the changes against is M140.

revidee committed:
feat: webrtc m140 support

note: this includes a similar fix as the PR336 from the mediasoup-client
repo which deals with webrtc changes and expectations for
payload-type/extension ids to stay consistent
note: some aestetics and clang-compile warnings have also been fixes, so
it compiles using cpp20/clang20

jmillan commented:
> I have to admit that no unit tests have been run, however, using it to produce/consume media works as expected in local tests.

Please, try to run the unit tests. The last PR updating libwebrtc was finally not merged due to it. I found those tests a PITA, but without them we are kind of lost everytime we make the slightest change.

I would not like to leave the project without unit tests.., if we are not able to keep them up to date we may be forced to remove them, which would in detriment of the project.

jmillan commented on src/PeerConnection.cpp:
```diff
@@ -102,6 +95,8 @@ namespace mediasoupclient
 			{
 				MSC_THROW_INVALID_STATE_ERROR("thread start errored");
 			}
+			auto trials = "WebRTC-SupportVP9SVC/EnabledByFlag_3SL3TL/";
```
Is this not enabled by default?

revidee committed:
fix: using function reference for getSendExtendedRtpCapabilities which can lead to lifetime issues

fix: update tests to use new webrtc m140 namespaces (rtc:: got removed)
fix: use webrtc::SdpType for Offer/Answers instead of mediasoups old
string types
fix: using extendedRtpCapabilities instead of RecvRtpCapabilities which
are expected by ortc::canReceive
change: link against single libwebrtc(.a)
change: use cpp20

revidee commented on src/mediasoupclient.cpp:
```diff
@@ -17,17 +14,14 @@ namespace mediasoupclient
 
 		MSC_DEBUG("mediasoupclient v%s", Version().c_str());
 
-		webrtc::field_trial::InitFieldTrialsFromString("WebRTC-SupportVP9SVC/EnabledByFlag_3SL3TL/");
```
@jmillan I moved the field trials init from being global to being local - since the global API got deprecated in WebRTC. Not sure if this is the default nowadays, I left it in to ensure compatibility.

revidee replied:
Just to follow up on this: The flag is - as far as I can tell - not included in WebRTC M140, so this should serve no function anymore (most likely it is indeed enabled by default now)

revidee commented:
So I ran the test and had to fix some minor issues:

- Some namespace issues in the test files that still used the removed `rtc::` namespace
- Using removed `mediasoup::SdpType::` types instead of the replacement from `webrtc::SdpType::`
- Using extendedRtpCapabilities instead of RecvRtpCapabilities which
are expected by `ortc::canReceive`, which otherwise leads to crashes due to invalid access of `preferredPayloadType`, which only exsist in the RecvRtpCaps
- I chose to link to a single static libwebrtc.a library since I think this is the most common webrtc compilation method - a single large static lib
- I bumped cpp standard to 20

**All tests are now passing.** However, I experience several issues while building:
- I had to use clang, when using gcc, multiple errors from webrtc include failed the built (maybe webrtc isn't compatible with gcc anymore? Not sure.)
- I used Chromium's libcxx and clang, since my webrtc built uses that and expects the target app to also use the chromium libcxx and optimally also use the chromium-supplied clang20
- I linked abseil 20250814.0 and openssl 3.3.2 manually, as well as several system libraries required by webrtc manually (they are not bundled in my static webrtc build in order to reduce code duplication in another app)

All in all, this was my cmake command:
```sh
cmake -DCMAKE_BUILD_TYPE=Debug -DLIBWEBRTC_INCLUDE_PATH=/home/.../include/webrtc \
-DLIBWEBRTC_BINARY_PATH=/home/.../lib -DMEDIASOUPCLIENT_BUILD_TESTS=ON \
-DMEDIASOUPCLIENT_LOG_DEV=ON -DMEDIASOUPCLIENT_LOG_TRACE=ON \
-DCMAKE_POLICY_VERSION_MINIMUM=3.14 -DCMAKE_EXPORT_COMPILE_COMMANDS=ON \
-DCMAKE_C_COMPILER="clang"   -DCMAKE_CXX_COMPILER="clang++"   \
-DCMAKE_CXX_FLAGS="-stdlib++-isystem${LIBCXX_ROOT}/include/c++/v1 -Wno-nullability-extension -Wno-nullability-completeness"   \
-DCMAKE_EXE_LINKER_FLAGS="-stdlib=libc++ -fuse-ld=lld -L${LIBCXX_ROOT}/lib -L/home/.../lib $(ls -A1 /home/.../abseil/.../lib | grep '\.a$' | sed 's/lib/-l/' | sed 's/\.a//' | tr '\n' ' ') -L/home/.../openssl/.../lib -lssl -lpthread -ldl -lcrypto -lz -lm -lXrandr -luuid -lXtst -lX11 -lrt -lexpat -lXdamage -lXcomposite -latomic -lgcc -lgobject-2.0 -lXfixes -lgbm -lgio-2.0 -lXext -lglib-2.0 -lXrender -ldrm -lgmodule-2.0 -Wl,-rpath,${LIBCXX_ROOT}/lib"  \
../
```
- `-DCMAKE_POLICY_VERSION_MINIMUM=3.14` was required since libsdptransfrom uses a pretty old min version which is not supported by default anymore by newer cmake versions (libsdptransform should raise the min cmake version)
- `-Wno-nullability-extension -Wno-nullability-completeness` is required (for clang) to mute several hundred warnings inside webrtc includes regarding _Nullable macros (and the mis-use of that, thanks google)
- `stdlib++-isystem${LIBCXX_ROOT}/include/c++/v1` is required only for my custom google libcxx
- `-fuse-ld=lld` is required for clang to use lld to read the static webrtc lib correctly, otherwise the arch is not recognized correctly during linking
- `$(ls -A1 /home/.../abseil/.../lib | grep '\.a$' | sed 's/lib/-l/' | sed 's/\.a//' | tr '\n' ' ')` links against all of abseil (which is usually bundled in webrtc)
- `-lssl -lpthread -ldl -lcrypto -lz -lm -lXrandr -luuid -lXtst -lX11 -lrt -lexpat -lXdamage -lXcomposite -latomic -lgcc -lgobject-2.0 -lXfixes -lgbm -lgio-2.0 -lXext -lglib-2.0 -lXrender -ldrm -lgmodule-2.0` is ripped from webrtc BUILD.gn files and links against the required system libs (and external openssl build)

jmillan commented:
Nice @revidee :+1:. Let me few days to come back into this.

jmillan commented:
Merging this. I'll create a new release. We'll need to work on enhancing the unit tests.

jmillan commented:
> using it to produce/consume media works as expected in local tests.

Can I ask what do you use for local testing?

revidee commented:
> > using it to produce/consume media works as expected in local tests.
> 
> Can I ask what do you use for local testing?



Sure. We're using a testing mediasoup server with a thin protocol layer and integrated libmediasoupclient into an existing P2P video app. We added the protocol support and switched from P2P to the SFU. With that we can spin up a mediasoup server + wrapper and test drive that locally.

It's unfortunately pretty deeply integrated into our app, it would be great for pure libmediasoupclient development to have a more stripped down app with minimal gui, if even.

jmillan commented:
version 3.5.0 released.

jmillan commented:
> It's unfortunately pretty deeply integrated into our app, it would be great for pure libmediasoupclient development to have a more stripped down app with minimal gui, if even.

Totally agreed. We have a demo [here](https://github.com/versatica/mediasoup-broadcaster-demo) that integrates with [mediasoup-demo](https://github.com/versatica/mediasoup-demo) which uses a [MediaStreamTrackFactory](https://github.com/versatica/mediasoup-broadcaster-demo/blob/v3/src/MediaStreamTrackFactory.cpp) which relies on libwebrtc test code.., which is very unfortunate as we have the same problems as for the unit test.

If we were only able to get rid of such libwebrtc test dependency I think we would be made for both unit tests and demo. Feel free to take a look and play around if you fancy, @revidee.

jmillan commented:
Website updated: https://mediasoup.org/documentation/v3/libmediasoupclient/installation/

revidee commented:
> Website updated: https://mediasoup.org/documentation/v3/libmediasoupclient/installation/


Great. However, I'm not sure if the linking args in the tests are correct for the "default"/wiki described way of building webrtc. I adjusted these for a single big static lib, but maybe the default args do produce many small .a libs for each TU.

We may want to double check that :)

jmillan commented:
> I'm not sure if the linking args in the tests are correct for the "default"/wiki described way of building webrtc.

Can you be more specific?

BTW: I've removed `mac_deployment_target` from `gn gen` command example.
